ex2_hal/iris/hardware_interface/source/iris.c

```diff
@@ -0,0 +1,460 @@
+                break;
+            }
+            case GET_DATA:
+            {
+                //TODO: write code to receive image count
+                break;
+            }
+            case FINISH:
+            {
+                ex2_log("Iris returns ACK on transfer image command");
+                return IRIS_HAL_OK;
+            }
+            case ERROR_STATE:
+            {
+                ex2_log("Iris returns NACK on transfer image command");
+                return IRIS_HAL_ERROR;
+            }
+        }
+    }
+    return IRIS_HAL_ERROR;
+}
+
+/**
+ * @brief
+ *   Sends a toggle sensors command to Iris
+ *
+ * @param[in] toggle
+ *   Flag to either turn on or off Iris sensors
+ *
+ * @return
+ *   Returns IRIS_HAL_OK if equipment handler returns IRIS_ACK, else IRIS_HAL_ERROR
+ **/
+IrisHALReturn iris_toggle_sensor_idle(IRIS_SENSOR_TOOGGLE toggle) {
+    IrisLowLevelReturn ret;
+
+    controller_state = SEND_COMMAND;
+
+    while (controller_state != FINISH && controller_state != ERROR_STATE) {
+        switch (controller_state) {
+            case SEND_COMMAND:
+            {
+                if (toggle == IRIS_SENSOR_ON) {
+                    ret = send_command(IRIS_ON_SENSOR_IDLE);
+                } else if (toggle == IRIS_SENSOR_OFF) {
+                    ret = send_command(IRIS_OFF_SENSOR_IDLE);
+                }
+                if (ret == IRIS_ACK) {
+                    controller_state = FINISH;
+                } else {
+                    controller_state = ERROR_STATE;
+                }
+                break;
+            }
+            case FINISH:
+            {
+                ex2_log("Iris returns ACK on toggling sensor command");
+                return IRIS_HAL_OK;
+            }
+            case ERROR_STATE:
+            {
+                ex2_log("Iris returns NACK on toggling sensor command");
+                return IRIS_HAL_ERROR;
+            }
+        }
+    }
+    return IRIS_HAL_ERROR;
+}
+
+/**
+ * @brief
+ *   Sends a housekeeping command to Iris and expects to receive housekeeping
+ *   statistics which then is stored into internal housekeeping struct
+ *
+ * @return
+ *   Returns a housekeeping data structure
+ **/
+IrisHALReturn iris_get_housekeeping(iris_housekeeping_data hk_data) {
+    IrisLowLevelReturn ret;
+
+    controller_state = SEND_COMMAND;
+
+    while (controller_state != FINISH && controller_state != ERROR_STATE) {
+        switch (controller_state) {
+            case SEND_COMMAND:
+            {
+                ret = send_command(IRIS_SEND_HOUSEKEEPING);
+                if (ret == IRIS_ACK) {
+                    vTaskDelay(1000);
+                    controller_state = GET_DATA;
+                } else {
+                    controller_state = FINISH;
+                }
+                break;
+            }
+            case GET_DATA:
+            {
+
+                uint16_t * housekeeping_buffer = (uint16_t*) pvPortMalloc(HOUSEKEEPING_SIZE * sizeof(uint16_t));
```

Comment:
Allocating an array of known size at compile time would be better done as a stack variable

Reply:
Changes made in this commit [951545a](https://github.com/AlbertaSat/ex2_obc_software/pull/189/commits/951545a2b30e874aa07abc3ef220393a6601d2b1)